### ISSUE TITLE
[ENH] Adding offset argument to transformers that do not support negative values

### ISSUE THREAD

**Issue by ngupta23** [alan-turing-institute/sktime#2211]
**Is your feature request related to a problem? Please describe.**
Many transformers do not support negative values (e.g. BoxCoxTransformer, LogTransformer)

**Describe the solution you'd like**
Similar to what is done for [ExponentTransformer](https://www.sktime.org/en/v0.10.1/api_reference/auto_generated/sktime.transformations.series.exponent.ExponentTransformer.html), would it be possible to add an offset argument for such transformers.

**Comment by ltsaprounis** [comment]
A valid alternative would be to create an offset transformer and use it in a pipeline with the other ones. 
Not sure what's best but makes sense to have a little debate on it.

**Comment by aiwalter** [comment]
you could use a `MinMaxScaler` before those transformers that can not deal with <0 or <=0 values, would be more accurate I guess.

**Comment by ngupta23** [comment]
Lots of options here, but it would be most convenient for users to have the `offset` argument and set it to "auto" as done in `ExponentTransformer`. That way users do not have to do anything additional. 

Of course, they could use the other alternatives that have been mentioned but it is more inconvenient to add additional wrappers. Just my 2 cents on this.

**Comment by Vasudeva-bit** [comment]
I feel it would be better (can skip refactoring other transformers which are supporting negative values eg: exponential) to add an `offset` argument to transformers that are not supporting negative values. And also, the log transformation of values between 0 and 1 (obtained from MinMax Scaler of default `feature_range=(0, 1)`) would give negative values that again need to be handled, I suppose. MinMax Scaler with `feature_range=(1, variable)` also seems to be a nice alternative to me. Please let me know your view, I am happy to know.

**Comment by fkiraly** [comment]
Hm, I tend to agree with @aiwalter - why not just use a `MinMaxScaler` as a pipeline element?
Adding `offset` to a million transformers seems like multiplication and a violation of the DRY principle.

**Comment by aiwalter** [comment]
other option I can think off is having an `OffsetTransformer` compostion similar to the `OptionalPassthrough`. The `OffsetTransformer` would hold the functionality to offset and possibly have params `offset: bool = True` and `value: float/int = 0`
.
So when you want to offset a transformer, you could do `OffsetTransformer(BoxCoxTransformer())`

**Comment by ngupta23** [comment]
>Adding offset to a million transformers seems like multiplication and a violation of the DRY principle.

@fkiraly Maybe having an `OffsetMixin` class would address this?

> other option I can think of is having an `OffsetTransformer` compostion similar to the `OptionalPassthrough`. The `OffsetTransformer` would hold the functionality to offset and possibly have params `offset: bool = True` and `value: float/int = 0`
> .
> So when you want to offset a transformer, you could do `OffsetTransformer(BoxCoxTransformer())`

Why is this better than just `BoxCoxTransformer()` with default argument `offset = "auto"`? It seems much more simpler from a users perspective so instead of tuning like `transformer__limits` to access pipeline params, one would only have to do `limits`. Plus less overhead when writing out the code for the user as well.

Maybe having both "offset" argument with a MixinClass and an `OffsetTransformer` might also be an option if there is no easy conclusion.

**Comment by aiwalter** [comment]
Yes I actually also dont prefer the composition. If we add it anyhow as a param we should also deprecate the `OptionalPassthrough` and have it in the same way as offset added to the transformers as it is also a functionality that we want to enable for all transformers. We could also add `offset` and `passthrough` params to the base class and give it to `super().__init__(offset, passthrough)`. The logic to passthrough/offset could be implemented in `fit()` and `transform()`, so we would only have to update `super` and the docstring for all transformers.

**Comment by fkiraly** [comment]
> @fkiraly Maybe having an OffsetMixin class would address this?

That wouldn't automatically add the offset to the parameters, so I think it wouldn't work.
It would have to be at least a decorator, and that just invites trouble...

> Why is this better than just BoxCoxTransformer() with default argument offset = "auto"? It seems much more simpler from a users perspective

Hm, I disagree - the user familiar with the Box-Cox transform willl probably expect the transformer to produce nan or to break if there are negative values for certain exponents. The "offsetting" logic will come as a surprise, and will conceptually not be part of someone's mental model of the Box-Cox transform!

User surprise and user frustrationn is the result - and in hindsight, I think that the "offset" logic does *not* really belong in the ` BoxCoxTransformer` either.

@ngupta23, what would be wrong about writing sth like

`OffsetNegatives() * BoxCoxTransformer()`, as @aiwalter proposes?

That makes it much clearer what is happening - design principle being, specification should map as closely as possible on the mental model.

**Comment by fkiraly** [comment]
> Yes I actually also dont prefer the composition. If we add it anyhow as a param we should also deprecate the `OptionalPassthrough` and have it in the same way as offset added to the transformers as it is also a functionality that we want to enable for all transformers.

Ok, I think that is a bad idea, @aiwalter. This will lead to monstrosities since all transformers now are going to contain composition with two other transformers. And why not add all the parameters of the min/max scaler too? That direction I really do not like. "Simple is better than complex."

**Comment by fkiraly** [comment]
So, I liked your first suggestion much better than the second, @aiwalter 😄

**Comment by ngupta23** [comment]
> User surprise and user frustrationn is the result - and in hindsight, I think that the "offset" logic does not really belong in the BoxCoxTransformer either.

@fkiraly - Did you mean does not belong in the `ExponentTransformer`? If so is it going to be removed?

> @ngupta23, what would be wrong about writing sth like
OffsetNegatives() * BoxCoxTransformer(), as @aiwalter proposes?

I am not familiar with this syntax. Are you referring to a nested pipeline like this?

```python
OffsetNegatives(BoxCoxTransformer())
```

**Comment by fkiraly** [comment]
> If so is it going to be removed?

I don't think so - it's already there, and deprecation is always a hassle.
There's a difference between *removing* sth and *not putting it there*.
I was just talking about a hypothetical "ideal state".

> I am not familiar with this syntax. Are you referring to a nested pipeline like this?

That's actual pipeline syntax coming in 0.11.0, you can chain elements with the `*` operator.

**Comment by aiwalter** [comment]
From design/conceptual perspective, I think having a `OffsetTransformer` look more reasonable, however I was just trying to think more from a user perspective. Having sth like `OffsetTransformer(OptionalPassthrough(BoxCox()))` is not so intuitive than having `BoxCox(passthrough, offset)` especially when it comes to tuning the params.

**Comment by ngupta23** [comment]
> From design/conceptual perspective, I think having a OffsetTransformer look more reasonable, however I was just trying to think more from a user perspective. Having sth like OffsetTransformer(OptionalPassthrough(BoxCox())) is not so intuitive than having BoxCox(passthrough, offset) especially when it comes to tuning the params.

Yes, that is exactly my point. The user has to add that to all transformers and then accessing pipeline parameters becomes more complicated. 

Plus from a users perspective, it is repetition if each transformer has to be wrapped in  2 additional transformers.
